docs/EXTENDED-ARCHIVES.md

```diff
@@ -0,0 +1,428 @@
+Etcher Extended Archives
+========================
+
+Etcher extended archives are archive files including an image and a special
+metadata directory including information that describes the image, device type,
+and vendor in detail, enabling Etcher to make use of such information to
+deliver advanced features and a more pleasant experience to the user.
+
+Entities
+--------
+
+### Vendor
+
+A vendor represents a manufacturer of a device type and potential publisher.
+
+*Some examples of image vendors are: Beagleboard, Raspberry Pi Foundation,
+Intel, and ODroid.*
+
+### Publisher
+
+A publisher represents an entity that maintains and publishes official images.
+
+*Some examples of image publishers are: Tizen, RetroPie, Resin.io, and CoreOS.*
+
+### Device
+
+A device represents a specific device type manufactured by a vendor.
+
+*Some examples of devices are: Beaglebone Black, Beaglebone Green, Intel
+Edison, and Raspberry Pi 3.*
+
+### Client
+
+A client represents the program making use of the information about the other
+entities.
+
+*At the moment of this writing, the only existent client is the Etcher application.*
+
+File format
+-----------
+
+The file format should have the following characteristics:
+
+- Streamability: The extended archive allows the client to consume all its
+  information in a streaming fashion
+
+- Compression: The extended archive should be compressed in order to save
+  network resources
+
+We base this format on the [`tar`][tar] archive format, which has the following
+benefits:
+
+- It's a nearly universal archive format (easy to find robust implementations)
+
+- It allows us to easily impose a file order, which greatly helps with
+  streamability. This is not trivial with `zip`, for instance
+
+- It precedes each file with a header construct, making it streaming friendly
+
+The [`tar`][tar] archive is then compressed with the [`xz`][xz] data
+compression format, which has the following benefits:
+
+- A native, robust, performant, cross-platform and streaming-based
+  decompression library for [node.js][nodejs], called
+  [`lzma-native`][lzma-native], which we battle tested at Etcher
+
+- A way to calculate the precise uncompressed size of the file no matter the
+  size (which is a [known limitation of `gzip` after 4
+  GB](https://github.com/resin-io/etcher/issues/638)), and an operation
+  [`lzma-native`][lzma-native] supports out of the box. This is not possible
+  with `bzip2`
+
+- Decent compression ratio, and compression/decompression time
+
+The resulting file should contain the `.etch` extension.
+
+Archive layout
+--------------
+
+The contents of an extended archive are the followings:
+
+- **REQUIRED** `.meta/manifest.json`
+- `.meta/publisher.svg`
+- `.meta/device.svg`
+- `.meta/image.svg`
+- `.meta/schema.json`
+- `.meta/image.bmap`
+- `.meta/instructions.markdown`
+- `.meta/release-notes.txt`
+- **REQUIRED** `image.<extension>`
+- `.meta/checksum`
```

Comment:
Oh, first I've heard of an "Etcher Server". I thought the idea was that it'd be Etcher itself that did the configuration (using reconfix), after the image had been written to disk and verified (with the original checksum) ?

AFAIK it won't be possible for Reconfix to work in a streaming fashion (since you can't know which parts of the filesystem will be modified), so you'll still need to finish the whole reconfix process before starting to stream the image. So there's no reason the checksum can't be included in the data (i.e. manifest.json) before the image itself?